### ISSUE TITLE
How to freeze parameters?

### ISSUE THREAD

**Issue by jatentaki** [deepmind/optax#290]
I'm looking for a way to freeze parameters in most efficient manner. One would be to take the current `TrainingState` (I'm coming from flax) and manually partition `state.params` like
```python3
raw_state: TrainingState = original_unfrozen_model
frozen_params, learnable_params = manually_partition_parameters(raw_state.params)

def frozen_apply_fn(params, *args, **kwargs):
  return raw_state.apply_fn(
    manually_merge_params(params, frozen_params),
    *args,
    **kwargs
  )

partially_frozen_state = TrainingState(
  params=learnable_params,
  apply_fn=frozen_apply_fn
)
```
and then create a new optimizer for `partially_frozen_state`. The downside of that approach is that it completely mutilates the structure of the pytree, making serialized states incompatible between the frozen and unfrozen variant, as well as requires manual splitting and merging of the parameters in a consistent manner (easy source of bugs).

An alternative would be to use `optax.masked` to just not apply updates to certain parameters. My question however is if JAX (XLA) is smart enough to optimize the computation of gradients w.r.t. masked-out parameters out of the compute graph. This isn't clear to me, because the general code flow is to create the neural network in flax, which returns the application function and a pytree of parameters, I then manually (without optax's involvement) compute the gradients w.r.t. all parameters, and only then those are passed to optax, which may use them or not. If the gradients for frozen parameters are computed but just discarded, this solution would fail to preserve GPU memory, a major reason for freezing the parameters in the first place. If that works, I'm perfectly satisfied with that approach.

I am submitting this as an issue, rather than a discussion, because currently searching the docs for "freeze"/"frozen" yields no hits at all, so the topic is either not covered at all or insufficiently discoverable.

**Comment by mkunesch** [comment]
Hi! Thanks a lot for the question and for raising the issue with documentation.

Freezing parameters is possible within optax by using a combination of `set_to_zero` and either `masked` or `multi_transform`. The `set_to_zero` is important since `masked` only keeps _updates_/_gradients_ unmodified by the gradient transformation, so the parameters would still be modified when non-zero updates are applied.

In terms of performance, I think the important point is whether the gradient calculations, optax gradient transformations, and the applications of updates to the parameters happen within the same `jax.jit` compiled function. Since `set_to_zero` uses `zeros_like` to set updates to zero, the calculation of gradients of frozen parts of the parameter tree should be dropped by `jax.jit`.

In terms of how we could make it easier for users to discover this, we could for example change the docstring of `set_to_zero` such that it also includes the phrase "freeze parameters" so that users searching for "freeze" can find it? Another option would be to add an alias for the combination of `set_to_zero` and `masked` and call it `freeze`, but I think the name would go against the optax model a bit (which is to transform gradients and those are set to zero rather than frozen). What do you think?

Thanks a lot again!

**Comment by rosshemsley** [comment]
Freezing is a common enough case that it might be nice to put an example under `Examples` in the online docs (https://optax.readthedocs.io/en/latest/).

I created a ticket for this at https://github.com/deepmind/optax/issues/296

**Comment by jatentaki** [comment]
@mkunesch sorry for the late reply, it slipped through. Thanks a lot for your explanation. I agree with @rosshemsley that an example would be the best, but it should also be searchable in the API reference itself. I think modifying the docstring to say something like "this operation can be used to freeze parameters as: <insert one line example>" would be good, adding explicity function name aliases shouldn't be necessary :)

**Comment by mkunesch** [comment]
Sounds good - I completely agree with all this! Thanks @rosshemsley for opening the issue with the example.

@jatentaki: would you like to file a PR for the docstring change or shall I?

**Comment by jatentaki** [comment]
I think I'd prefer you do this. I haven't had the time to actually try this solution yet, so I would be writing about something I haven't experienced myself. If that's not a problem, please file the PR, otherwise I can do that after my current deadlines...

**Comment by mkunesch** [comment]
Sounds good, I'm more than happy to do it! I just wanted to make sure we don't both work on it at the same time.

**Comment by davisyoshida** [comment]
There's one aspect of the `set_to_zero()` solution which is a bit not-nice. Specifically, anything which uses the gradients downstream will cause constant folding to happen.

In my case, it turned out that the set of steps needed for dynamic loss scaling (checking whether gradients are finite, rescaling) led to extremely long JIT times (maybe half an hour?) when freezing a large number of parameters in LLaMA-7B.

It would be really nice if optax could add support for some sort zero type, like `jax.ad.Zero` or maybe JAX's `float0` dtype. In terms of optax itself, I think the minimal change to support this would be a tweak to `apply_updates()`:

```python
def apply_updates(params: base.Params, updates: base.Updates) -> base.Params:
  return jax.tree_util.tree_map(
      lambda p, u: jnp.asarray(p + u).astype(jnp.asarray(p).dtype) if not is_const_zero(u) else p,
      params, updates)
```

This is backwards compatible, and then client code would be able to branch on whether a gradient was this Zero type for things like `jmp.all_finite`.

**Comment by carlosgmartin** [comment]
Note that there's now [optax.freeze](https://optax.readthedocs.io/en/latest/api/transformations.html#optax.freeze), as described in [this tutorial](https://optax.readthedocs.io/en/latest/_collections/examples/freezing_parameters.html).
